WebCrypto X25519 algorithm docs (#35441)
mdn/content: files/en-us/web/api/subtlecrypto/derivekey/index.md

#### X25519

X25519 is a key agreement algorithm like ECDH, but built on the [Curve25519](https://en.wikipedia.org/wiki/Curve25519) elliptic curve, which is part of the Edwards-Curve Digital Signature Algorithm (EdDSA) family of algorithms defined in {{rfc("8032")}}.

The Curve25519 algorithms are widely used in cryptography, and are considered to be some of the most efficient/fast available.

> By comparison with the NIST (National Institute of Standards and Technology) curve key exchange algorithms used with ECDH, Curve25519 is simpler to implement, and its non-governmental origin means that the decisions behind its design choices are transparent and open.

Author: Figured it is worth adding a bit more background. This summary is synthesized from the bug report https://bugzilla.mozilla.org/show_bug.cgi?id=1904836#c4 and wikipedia. Essentially it is faster  than the NIST algorithms, and because it is easier to implement, less likely to be broken by a poor implementation.

The parameters used for the NIST algorithms were also selected through an opaque "government" process and are less trusted by some people that the Curve25519 algorithms, for which the decision making process is perceived to be fully open.

}
```

> ### X25519 derived shared key

Author: @wbamberg Because this is a live example and above the algorithm description has the same name.

I'll fix this by giving all of the examples a more expansive heading. Linter giving me trouble with this right now (it thinks because I rename one of the headings I have broken the links to it)

> ### X25519 derived shared key

Author: Fixed. Thanks for the `--no-verify` tip.

This example is much the same as the example above, except that it uses keys generated using the X25519 algorithm rather than ECDH, and it is a live example.

In this example Alice and Bob each generate an X25519 key pair, then exchange public keys.
They then each use `deriveKey()` to derive a shared AES key from their own private key and each other's public key.
Their respective AES keys can encrypt and decrypt messages send between them.

#### HTML

> First we define an HTML {{HTMLElement("input")}} for user to enter the plaintext message that Alice will send.

Author: > (Although this is a bit weird because usually Alice is is considered "the user".)

Good point. I think I will make that a bit less weird like this:

```suggestion
First we define an HTML {{HTMLElement("input")}} you to enter the plaintext message that "Alice" will send.
```

> This is followed by another two elements for displaying the cyphertext after Alice has encypted the plain text with her derived private key, and for displaying the text after Bob has decrypted it with his derived private key.

Author: Thank you - I was assuming "private key" could be used for any key that isn't to be shared. Do you think it easonable to update https://developer.mozilla.org/en-US/docs/Glossary/Key with the definition and link from this page?

> This is followed by another two elements for displaying the cyphertext after Alice has encypted the plain text with her derived private key, and for displaying the text after Bob has decrypted it with his derived private key.

Reviewer: You mean something like:

"In [symmetric-key cryptography](https://developer.mozilla.org/en-US/docs/Glossary/Symmetric-key_cryptography), the same key is used for both encryption and decryption."

->

"In [symmetric-key cryptography](https://developer.mozilla.org/en-US/docs/Glossary/Symmetric-key_cryptography), the same key is used for both encryption and decryptio. This key is known as the _secret key_."

I would also have a para break before the bit about public-key crypto.

> link from this page?

I don't know about that. It's a very generic glossary page and we are deep in the weeds here. If we want this para to be absolutely clear, it would be better to so say "AES key" rather than "secret key".

> This is followed by another two elements for displaying the cyphertext after Alice has encypted the plain text with her derived private key, and for displaying the text after Bob has decrypted it with his derived private key.

Author: Yes, that is exactly what I mean, though I used the words:

> This key is referred to as the _shared key_, or the _secret key_.

Feel free to directly subedit i you don't like that.

At the end is a button that the user can click to start the encryption process.

```html
<label for="message">Plaintext message from Alice (Enter):</label>
<input
  type="text"
  id="message"
  name="message"
  size="50"
  value="The lion roars near dawn" />

<label for="encrypted">Encrypted (Alice)</label>
<input
  type="text"
  id="encrypted"
  name="encrypted"
  size="25"
  value=""
  readonly />

<label for="decrypted">Decrypted (Bob)</label>
<input
  type="text"
  id="decrypted"
  name="decrypted"
  size="50"
  value=""
  readonly />

<input id="encrypt-button" type="button" value="Encrypt" />
```

```html hidden
<pre id="log"></pre>
```

```css hidden
input {
  display: block;
}
#log {
  height: 150px;
  width: 90%;
  white-space: pre-wrap; /* wrap pre blocks */
  overflow-wrap: break-word; /* break on words */
  overflow-y: auto;
  padding: 0.5rem;

>   border: 1px solid black;

Reviewer: This could use some margins to group elements: it's very hard to read as is.

}
```

#### JavaScript

```js hidden
const logElement = document.querySelector("#log");
function log(text) {
  logElement.innerText = `${logElement.innerText}${text}\n`;
  logElement.scrollTop = logElement.scrollHeight;
}
```

The code below shows how we use `deriveKey()`.

> This takes the algorithm and public key of the remote party, the private key of the local party, they algorithm of the derived key, and the "usages" for the new key, and creates a derived key.

Reviewer: This has a couple of issues. It's a bit hard to read because of the multiple "and": "they algorithm of the derived key, and the "usages" for the new key, and creates a derived key" so it would be better split into two sentences, one for params and one for return value. Also, though, it reads like an exhaustive list, but then omits the `false` parameter. Also an a exhaustive list seem like just rehashing the API reference up the page. Maybe it would be better if we described the parameters in more concrete terms:

```suggestion
We pass in the remote party's X25519 public key, the local party's X25519 private key, and specify that the derived key should be an AES-GCM key. We also set the derived key to be non-extractable, and suitable for encryption and decryption.

The function returns the derived AES-GCM key.
```

Author: > Also an a exhaustive list seem like just rehashing the API reference up the page. Maybe it would be better if we described the parameters in more concrete terms:

Absolutely. Much better.

We use this further down in the code to create shared keys for Bob and Alice.

```js
/*
Derive an AES key, given:
- our X25519 private key
- their X25519 public key
*/
function deriveSecretKey(privateKey, publicKey) {
  return window.crypto.subtle.deriveKey(
    {
      name: "X25519",
      public: publicKey,
    },
    privateKey,
    {
      name: "AES-GCM",
      length: 256,
    },
    false,
    ["encrypt", "decrypt"],
  );
}
```

> Next we define the functions that Alice will use to UTF-8 encode and then encrypt her plaintext message, and that Bob will use to decrypt and then decode the message.

Reviewer: ```suggestion
Next we define the functions that Alice will use to {{glossary("UTF-8")}} encode and then encrypt her plaintext message, and that Bob will use to decrypt and then decode the message.
```

> The both take the shared key of the caller, an "initialization vector", and the text to be encrypted or decrypted.
> 
> The [AES-GCM](/en-US/docs/Web/API/SubtleCrypto/encrypt#aes-gcm) algorithm requires that the initialization vector passed to the `encrypt()` and `decrypt()` methods is the same.
> Note however that while it has to be unique, it does not have to be secret, and can be shared with the remote party along with the public keys (in this case we'll just make it directly available).
> For more information see {{domxref("AesGcmParams")}}, which describes the object passed to the methods.

Reviewer: ```suggestion
They both take as arguments the shared AES key, an [initialization vector](/en-US/docs/Web/API/AesGcmParams#iv), and the text to be encrypted or decrypted.

The same initialization vector must be used for encryption and decryption, but it does not need to be secret, so usually it is sent alongside the encrypted message. In this case, though, since we're not actually sending a message, we just make it directly available.
```

I don't think the quotes on "initialization vector" are helpful here, and we do have a good place to link. Also we would _not_ generally shared the IV "along with the public keys", because typically we would only share the public key once, then generate one (or more) secret keys, and use the shared to to encrypt more than one message. whereas IVs must not be reused for more than one message.

Author: Yes, this was rubbish. Thank you.

```js
async function encryptMessage(key, initializationVector, message) {
  try {
    const encoder = new TextEncoder();
    encodedMessage = encoder.encode(message);
    // iv will be needed for decryption
    return await window.crypto.subtle.encrypt(
      { name: "AES-GCM", iv: initializationVector },
      key,
      encodedMessage,
    );
  } catch (e) {
    console.log(e);
    return `Encoding error`;
  }
}

async function decryptMessage(key, initializationVector, ciphertext) {
  try {
    const decryptedText = await window.crypto.subtle.decrypt(
      // The iv value must be the same as that used for encryption
      { name: "AES-GCM", iv: initializationVector },
      key,
      ciphertext,
    );

    const utf8Decoder = new TextDecoder();
    return utf8Decoder.decode(decryptedText);
  } catch (e) {
    console.log(e);
    return "Decryption error";
  }
}
```

The `agreeSharedSecretKey()` function below is called on loading to generate pairs and shared keys for Alice and Bob.
It also adds a click handler for the "Encrypt" button that will trigger encryption and then decryption of the text defined in the first `<input>`.

> Note that alll the code is inside a `try...catch` handler, to ensure that we can log the case where key generation fails because the X25519 algorithm is not supported.

Reviewer: ```suggestion
Note that all the code is inside a `try...catch` handler, to ensure that we can log the case where key generation fails because the X25519 algorithm is not supported.
```

    const alicesSecretKey = await deriveSecretKey(
      alicesKeyPair.privateKey,
      bobsKeyPair.publicKey,
    );

    log(
      `alicesSecretKey: ${alicesSecretKey.type} (algorithm: ${JSON.stringify(
        alicesSecretKey.algorithm,
      )}, usages: ${alicesSecretKey.usages}), `,
    );

    // Bob generates the same secret key using his private key and Alice's public key.
    const bobsSecretKey = await deriveSecretKey(
      bobsKeyPair.privateKey,
      alicesKeyPair.publicKey,
    );

    log(
      `bobsSecretKey: ${bobsSecretKey.type} (algorithm: ${JSON.stringify(
        bobsSecretKey.algorithm,
      )}, usages: ${bobsSecretKey.usages}), `,
    );

    // Get access for the encrypt button and the three inputs
    const encryptButton = document.querySelector("#encrypt-button");
    const messageInput = document.querySelector("#message");
    const encryptedInput = document.querySelector("#encrypted");
    const decryptedInput = document.querySelector("#decrypted");

    // Define the initialization vector that will be used for encrypting and decrypting the message.
    const initializationVector = window.crypto.getRandomValues(
      new Uint8Array(8),

>     );

Reviewer: IV must not be used for more than one message. So this must be created anew in each encrypt call, as ECDH does: https://github.com/mdn/dom-examples/blob/08dadaa1f2447d7412c9c8488a304f28b53dfadf/web-crypto/derive-key/ecdh.js#L26.

Author: Fixed by moving it into the click handler. Thank you.

      // Alice can use her copy of the shared key to encrypt the message.
      const encryptedMessage = await encryptMessage(
        alicesSecretKey,
        initializationVector,
        messageInput.value,
      );

      // We then display part of the encrypted buffer and log the encrypted message
      let buffer = new Uint8Array(encryptedMessage, 0, 5);
      encryptedInput.value = `${buffer}...[${encryptedMessage.byteLength} bytes total]`;

      log(
        `encryptedMessage: ${buffer}...[${encryptedMessage.byteLength} bytes total]`,
      );

      // Bob uses his shared secret key to decrypt the message.
      const decryptedCypertext = await decryptMessage(
        bobsSecretKey,
        initializationVector,
        encryptedMessage,
      );

      decryptedInput.value = decryptedCypertext;
      log(`decryptedCypertext: ${decryptedCypertext}`);
    });
  } catch (e) {
    log(e);
  }
}

// Finally we call the method to set the example running.
agreeSharedSecretKey();
```

#### Result

Press the Encrypt button to encrypt the text in the top `<input>` element, displaying the encrypted cypertext and decrypted cyphertext in the following two elements.

> The log area at the bottom provide information about the keys that are generated by the code.

Reviewer: Could we have a blank line added after each "Encrypt" click? Otherwise it gets really hard to read the log output if you click more than once:

```
Created Alices's key pair: (algorithm: {"name":"X25519"}, usages: deriveKey)
Created Bob's key pair: (algorithm: {"name":"X25519"}, usages: deriveKey)
alicesSecretKey: secret (algorithm: {"name":"AES-GCM","length":256}, usages: encrypt,decrypt), 
bobsSecretKey: secret (algorithm: {"name":"AES-GCM","length":256}, usages: encrypt,decrypt), 
Plaintext: The lion roars near dawn
encryptedMessage: 203,106,123,194,253...[40 bytes total]
decryptedCypertext: The lion roars near dawn
Plaintext: The lion roars near dawn
encryptedMessage: 203,106,123,194,253...[40 bytes total]
decryptedCypertext: The lion roars near dawn
```

    type="text"
    id="encrypted"
    name="encrypted"
    size="30"
    value=""
    readonly />

  <label for="results">Decrypted (Bob)</label>
  <input
    type="text"
    id="decrypted"
    name="decrypted"
    size="50"
    value=""
    readonly />
</div>
```

```html hidden
<pre id="log"></pre>
```

```css hidden
input {
  display: block;
  margin: 5px 0px 5px 0px;
}
#results {
  margin-top: 20px;
}

#log {
  height: 150px;
  width: 90%;
  white-space: pre-wrap; /* wrap pre blocks */
  overflow-wrap: break-word; /* break on words */
  overflow-y: auto;
  padding: 0.5rem;
  border: 1px solid black;
  margin-top: 20px;
}
```

#### JavaScript

```js hidden
const logElement = document.querySelector("#log");
function log(text) {
  logElement.innerText = `${logElement.innerText}${text}\n`;
  logElement.scrollTop = logElement.scrollHeight;
}
```

The code below shows how we use `deriveKey()`.
We pass in the remote party's X25519 public key, the local party's X25519 private key, and specify that the derived key should be an AES-GCM key.
We also set the derived key to be non-extractable, and suitable for encryption and decryption.

The function returns the derived AES-GCM key.

> We use this further down in the code to create shared keys for Bob and Alice.

Author: Ah, the "this" is ambiguous. This is supposed to be "We use this _function_ further down ..."

> We use this further down in the code to create shared keys for Bob and Alice.

Reviewer: Oh sorry, I understand now. But yes, it is ambiguous.

      // This must be regenerated for every message!
      const initializationVector = window.crypto.getRandomValues(
        new Uint8Array(8),
      );

      // Alice can use her copy of the shared key to encrypt the message.
      const encryptedMessage = await encryptMessage(
        alicesSecretKey,
        initializationVector,
        messageInput.value,
      );

      // We then display part of the encrypted buffer and log the encrypted message
      let buffer = new Uint8Array(encryptedMessage, 0, 5);
      encryptedInput.value = `${buffer}...[${encryptedMessage.byteLength} bytes total]`;

      log(
        `encryptedMessage: ${buffer}...[${encryptedMessage.byteLength} bytes total]`,
      );

      // Bob uses his shared secret key to decrypt the message.
      const decryptedCypertext = await decryptMessage(
        bobsSecretKey,
        initializationVector,
        encryptedMessage,
      );

      decryptedInput.value = decryptedCypertext;
      log(`decryptedCypertext: ${decryptedCypertext}\n`);
    });
  } catch (e) {
    log(e);
  }
}

// Finally we call the method to set the example running.
agreeSharedSecretKey();
```

#### Result

> Press the Encrypt button to encrypt the text in the top `<input>` element, displaying the encrypted cypertext and decrypted cyphertext in the following two elements.

Reviewer 2: `cypher` is [british](https://www.merriam-webster.com/dictionary/cypher).

> Press the Encrypt button to encrypt the text in the top `<input>` element, displaying the encrypted cypertext and decrypted cyphertext in the following two elements.

Author: Also Australian :-(
